sql-parser/src/main/java/io/crate/sql/tree/FrameBound.java

```diff
@@ -23,15 +23,134 @@
 package io.crate.sql.tree;
 
 import javax.annotation.Nullable;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
 
 public class FrameBound extends Node {
 
     public enum Type {
-        UNBOUNDED_PRECEDING,
-        PRECEDING,
-        CURRENT_ROW,
-        FOLLOWING,
-        UNBOUNDED_FOLLOWING
+        UNBOUNDED_PRECEDING {
+            @Override
+            public int getStart(int pStart,
+                                int pEnd,
+                                int currentFrameStartIdx,
+                                int currentRowIdx,
+                                boolean isOrdered,
+                                BiPredicate<Integer, Integer> arePeers) {
+                return pStart;
+            }
+
+            @Override
+            public int getEnd(int pStart,
+                              int pEnd,
+                              int currentRowIdx,
+                              boolean isOrdered,
+                              BiFunction<Integer, Integer, Integer> findFirstNonPeer) {
+                throw new IllegalStateException("UNBOUNDED PRECEDING cannot be the start of a frame");
+            }
+        },
+        PRECEDING {
+            @Override
+            public int getStart(int pStart,
+                                int pEnd,
+                                int currentFrameStartIdx,
+                                int currentRowIdx,
+                                boolean isOrdered,
+                                BiPredicate<Integer, Integer> arePeers) {
+                throw new UnsupportedOperationException("Custom PRECEDING frames are not supported");
+            }
+
+            @Override
+            public int getEnd(int pStart,
+                              int pEnd,
+                              int currentRowIdx,
+                              boolean isOrdered,
+                              BiFunction<Integer, Integer, Integer> findFirstNonPeer) {
+                throw new UnsupportedOperationException("Custom PRECEDING frames are not supported");
+            }
+        },
+        CURRENT_ROW {
+            @Override
+            public int getStart(int pStart,
+                                int pEnd,
+                                int currentFrameStartIdx,
+                                int currentRowIdx,
+                                boolean isOrdered,
+                                BiPredicate<Integer, Integer> arePeers) {
```

Comment:
I'm wondering if we should opt to pass in the sortedRows + the comparator instead of `isOrdered` and predicates/bi-functions. 
I've the feeling that we're now abstracting out something that could very well be in the responsibility of the frame types.

I think it could even use the `findFirstNonPeer` concretely. (And it could be moved into some kind of collections class, or maybe within `Lists2`)

If you've a fairly strong opinion that it is cleaner with this abstraction in place it is also fine for me to keep it.

Reply:
I bounced back between this and what I proposed several times yesterday but the simplification of testing made me decide to keep the rows and comparator out. I'll sketch this other way in another commit and we can take it from there and decide which one's better.